src/main/java/io/jenkins/lib/versionnumber/JavaSpecificationVersion.java

```diff
@@ -0,0 +1,95 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package io.jenkins.lib.versionnumber;
+
+import hudson.util.VersionNumber;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * Java Version Specification.
+ * Implemented according to https://openjdk.java.net/jeps/223
+ * @author Oleg Nenashev
+ * @since 1.6
+ */
+public class JavaSpecificationVersion extends VersionNumber {
+
+    private static final String JAVA_SPEC_VERSION_PROPERTY_NAME = "java.specification.version";
+
+    public static final JavaSpecificationVersion JAVA_5 = new JavaSpecificationVersion("1.5");
+    public static final JavaSpecificationVersion JAVA_6 = new JavaSpecificationVersion("1.6");
+    public static final JavaSpecificationVersion JAVA_7 = new JavaSpecificationVersion("1.7");
+    public static final JavaSpecificationVersion JAVA_8 = new JavaSpecificationVersion("1.8");
+    public static final JavaSpecificationVersion JAVA_9 = new JavaSpecificationVersion("9");
+    public static final JavaSpecificationVersion JAVA_10 = new JavaSpecificationVersion("10");
+    public static final JavaSpecificationVersion JAVA_11 = new JavaSpecificationVersion("11");
+    public static final JavaSpecificationVersion JAVA_12 = new JavaSpecificationVersion("12");
+
+    /**
+     * Constructor which automatically normalizes version strings.
+     * @param version Java specification version, should follow JEP-223 or the previous format.
+     * @throws NumberFormatException Illegal Java specification version number
+     */
+    public JavaSpecificationVersion(@Nonnull String version)
+            throws NumberFormatException {
+        super(normalizeVersion(version));
+    }
+
+    @Nonnull
+    private static String normalizeVersion(@Nonnull String input)
+            throws NumberFormatException {
+        input = input.trim();
+        if (input.startsWith("1.")) {
+            String[] split = input.split("\\.");
+            if (split.length != 2) {
+                throw new NumberFormatException("Malformed old Java Specification Version. " +
+                        "There should be exactly one dot and something after it: " + input);
+            }
+            input = split[1];
+        }
+
+        int majorVersion = Integer.parseInt(input);
+        if (majorVersion > 8) {
+            return input;
+        } else {
+            return "1." + input;
+        }
+    }
+
+    /**
+     * Get the Java Specification version for the current JVM
+     * @return Java Specification version or {@code null}
+     * @throws NumberFormatException Version parsing error
+     * @throws IllegalArgumentException JVM does not specify the mandatory {@link #JAVA_SPEC_VERSION_PROPERTY_NAME} property.
```

Comment:
It's not the exception raised: `throw new IllegalStateException(`

Reply:
Is it worth to mention the `NumberFormatException` as well if the property is bad defined?